agent/config/config.go

```diff
@@ -353,6 +353,8 @@ func environmentConfig() (Config, error) {
 			errs = append(errs, err)
 		}
 	}
+	containerMetadataEnabled := utils.ParseBool(os.Getenv("ECS_ENABLE_CONTAINER_METADATA"), false)
+	dataDirOnHost := os.Getenv("ECS_HOST_DATA_DIR")
```

Comment:
What happens when ECS Init/whatever's starting the Agent mounts some other directory as Agent's data directory (`-v /tmp:/data`) sets `ECS_HOST_DATA_DIR` to `"/var/lib/ecs"`

It doesn't seem like a strong enough abstraction to be dependent on Agent configuration options to expect `ECS_HOST_DATA_DIR` to be the same as whatever's being mounted as `/data` on the host. My main concern is the late-binding/asynchronous failures this could lead to if the wrong host mount was specified. Wondering if there are better alternatives here. Can we at the very least inspect the Agent container and check if `ECS_HOST_DATA_DIR` is mounted?

Reply:
I think we can do away with `ECS_HOST_DATA_DIR` if we directly inspect the Agent container for whichever directory is mounted as `/data`.

Reply:
hm you're right, there's a failure case here. i agree it makes sense to validate ECS_HOST_DATA_DIR is mounted by inspecting the Agent container.

however, i'm not sure if it makes more sense to fail during agent creation with a configuration error since the user will be expecting the metadata feature to be enabled, or if it would make more sense for us to surface a warning log and just disable the metadata feature.

i think failing explicitly during agent creation, `newAgent(...)`, makes more sense since this would be a configuration error. whereas logging, disabling the feature and moving forward seems like a poor experience since the only ways for the user to find out about these errors would be to catch it in the log, or through containers failing because they expect a metadata file.

Reply:
> i think failing explicitly during agent creation, newAgent(...), makes more sense since this would be a configuration error.

I agree

Reply:
@aaithal so as per our discussion earlier, we will call this out in the README under usage instructions for the metadata feature.

Reply:
There are a few assumptions here that I think are problematic:

* Assuming that `ECS_HOST_DATA_DIR` is `/var/lib/ecs` on the host
* Assuming that the ECS agent is running inside a container (it runs *outside* a container on Windows)
* Assuming that the ECS agent container has the name `ecs-agent` so you can inspect the container (true on ECS-optimized AMIs and other places where `ecs-init` is used, but not elsewhere)
* Assuming that you can get the ECS agent container ID through its hostname so you can inspect the container (won't be true if the UTS namespace is shared)

Reply:
Adding to samuelkarp's second point, I believe the metadata feature does not work when the agent is run outside a container on Linux environments. Initially I was told this is a minor concern, but do we have use use cases for the agent being run outside a a container on Linux? For Windows, this situation is explicitly handled (As noted in the current documentation).